### DISCUSSION TITLE
Z-Stack Data Storage and Image Quality

### DISCUSSION THREAD

**Discussion by jmdelahanty** [NeurodataWithoutBorders/helpdesk#4] [Q&A] [answered]
Hey everyone!

Do you have any advice on how many frames is adequate for a reasonable image quality in a z-stack for 2-photon images? Grabbing single images for a plane just looks like noise to me so I was planning on collecting 8 or 16 stacks individually and averaging them together.

Thanks for the tips and advice!

**Comment by bendichter** [comment, marked as answer]
@jmdelahanty It might be OK to average frames together as part of a preprocessing program, but I'd suggest storing all of the raw data to start with, even if individual frames look noisy to the eye.

**Reply by jmdelahanty** [reply]
Hey Ben! Thanks for the advice, I'll be sure to do that. To use the grayscale volume extension correctly, would each frame be it's own ophys series or can each stack be one series where just the z-dimension changes? And related to your answer, the averaged together frames can be its own series in the same file right?

Since these stacks will be affiliated with sessions where we record some live neural activity, can the stacks be part of the same NWB file? So there would be the raw neural recording data, 8-16 raw stacks, raw subject behavior data, and a raw video of the subject's face to start. Then I'll be appending processed data to the NWB file as it goes through the pipeline until the processing is done. I'm worried about how large the files might get if I include all this raw and processed information into the file. I don't want my processing pipelines to get slowed down after each set of data advances because the file is getting bigger and bigger. This might be a dumb question, but when reading a subset of the HDF5 file (i.e. one optical channel) is the entire file loaded first and then the relevant selection is loaded or does HDF5 only load the data you specify from disk?

**Reply by oruebel** [reply]
>  but when reading a subset of the HDF5 file (i.e. one optical channel) is the entire file loaded first and then the relevant selection is loaded or does HDF5 only load the data you specify from disk?

Both PyNWB and MatNWB use lazy data load, i.e., only the base organization of the data and small metadata is loaded on read. The bulk data is then loaded on request when you slice into the arrays.

> Then I'll be appending processed data to the NWB file as it goes through the pipeline until the processing is done. I'm worried about how large the files might get if I include all this raw and processed information into the file.

If file size is an issue, then you can also save the results from all stages into their own NWB file. Using external links you could then link the files together to be able to open the different parts as if they were stored in one file. To keep things simple, it might be easiest to start with a workflow where you write everything into one file and if that causes problems update the workflow to use multiple files. For a short tutorial on how to link files see the tutorial here https://pynwb.readthedocs.io/en/stable/tutorials/advanced_io/linking_data.html

**Reply by jmdelahanty** [reply]
Thanks for the response Oliver! That makes sense re lazy loading, thanks for the explanation.

> To keep things simple, it might be easiest to start with a workflow where you write everything into one file and if that causes problems update the workflow to use multiple files.

That makes sense and that's what I'll try to do. Do you have any rules of thumb for file size best practices based on your experience? As in when should I start to worry about size?

**Reply by oruebel** [reply]
> Do you have any rules of thumb for file size best practices based on your experience? As in when should I start to worry about size?

As a general rule, I would say, if you know the data is something that you want to keep (and share) then placing it in the file is fine. If the data is something you believe you may not want to delete, e.g., some preliminary analysis, then I would place it in a separate file so that you can easily remove it. In terms of file-size, it really depends on what you expect to do with the data.  On AWS and Azure clouds the file size limit is around 5TB (I think), so I would not go larger than that. The main issues I've encountered with large files is that transferring the data becomes more challenging, in particular when moving from a high performance system (e.g,. cloud or supercomputer) to a less performant system (e.g., laptop or backup tape store) with limited network bandwidth or storage capacity. Rather than strictly worrying about size, I would ask the question "How do you typically use the data?" E.g,:
- If the data belongs together and you usually want to be able to use it all together, then placing it into one file is fine. 
- If all the data is typically used in a write-once-read-many type of pattern (i.e., where you generate the data and then not modify it again) then placing it in the same file is fine.
- If some data is static but other parts are modified often (or need to be deleted) then I would place the parts that change in a separate file

**Reply by jmdelahanty** [reply]
Thank you so much for all the clarification Oliver! This all makes sense. Thankfully, we'll not be getting anywhere close to 5TB (at least I'm pretty darn sure we won't!). My plan is to have all the data in the file be static once we figure out how to optimally process our fluorescence microscopy data with Suite2P. From there people could read in the data on their local machines and do stats/etc. I'm hoping that the NWB file is something that users of the system basically don't have to build/interact with beyond loading it. Ideally, they won't write to it either since the NWB file should be a complete representation of the given recording session and writing to it might make it impossible for other users to load and analyze.

**Reply by oruebel** [reply]
> writing to it might make it impossible for other users to load and analyze

when the HDF5 file opened normally (i.e., with 'w' or 'a') that is correct. HDF5 has support for "single writer multiple reader" access as well but that I would consider that to be an advanced use case (https://docs.h5py.org/en/latest/swmr.html#single-writer-multiple-reader-swmr)

**Reply by jmdelahanty** [reply]
Do you think that it's a good idea to restrict write access? In other words, is it best practice to basically have a pipeline that produces a complete, static NWB file that users then analyze or should I try to see how flexible the NWB file for a session should be? The only thing that I'm worried about is that since using NWB is totally new to the lab that when we're analyzing it we might accidentally ruin it when running things on it.

Edit: Maybe it's better to say that my goal for the pipeline is to automatically process data to an analyzable state without any end user input/handholding required so people can focus on their science and not their data formatting. I volunteered in a lab at Davis that basically had an automatic pipeline that would let you know when an animal's data for a session was ready for analysis and they had premade scripts that could be called on any file they produced and let them visualize it, analyze it, and interact with it. It's a dream to do that here!

**Reply by oruebel** [reply]
> Do you think that it's a good idea to restrict write access? In other words, is it best practice to basically have a pipeline that produces a complete, static NWB file that users then analyze or should I try to see how flexible the NWB file for a session should be?

In general, I think that's probably a good rule of thumb, at least for all data that you want to be static (e.g., for raw recordings and data from standardized pipelines). For analyses (which tend to change), saving results to separate NWB is probably preferred, at least until you are confident that the analysis results are good and are what you want to share.
